tests/test_models.py

```diff
@@ -236,9 +247,124 @@ def test_find_or_404_found(self):
+        
+    def test_create_a_product(self):
+        """ Create a Product and assert that it exists """
+        product = self._create_product()
+        self.assertTrue(product != None)
+        self.assertEqual(product.id, None)
+        self.assertEqual(product.name, "Macbook")
+
+    def test_add_a_product(self):
+        """ Create a Product and add it to the database """
+        products = Product.all()
+        self.assertEqual(products, [])
+
+        product = self._create_product()
+        self.assertTrue(product != None)
+        self.assertEqual(product.id, None)
+        product.create()
+
+        # Assert that it was assigned an id and shows up in the database
+        self.assertNotEqual(product.id, None)
+        products = Product.all()
+        self.assertEqual(len(products), 1)
+
+    def test_update_a_product(self):
+        """ Update a Product """
+        product = self._create_product()
+        logging.debug(product)
+        product.create()
+        logging.debug(product)
+        self.assertEqual(product.id, 1)
+        # Change it an save it
+        product.name = "Updated Name"
+        original_id = product.id
+        product.save()
+        self.assertEqual(product.id, original_id)
+        self.assertEqual(product.name, "Updated Name")
+        # Fetch it back and make sure the id hasn't changed
+        # but the data did change
+        products = Product.all()
+        self.assertEqual(len(products), 1)
+        self.assertEqual(products[0].id, 1)
+        self.assertEqual(products[0].name, "Updated Name")
+
+    def test_delete_a_product(self):
+        """ Delete a Product """
+        product = self._create_product()
+        product.create()
+        self.assertEqual(len(Product.all()), 1)
+        # delete the product and make sure it isn't in the database
+        product.delete()
+        self.assertEqual(len(Product.all()), 0)
+
+    def test_serialize_a_product(self):
+        """ Test serialization of a Product """
+        product = self._create_product()
+        data = product.serialize()
+        self.assertNotEqual(data, None)
+        self.assertIn("id", data)
+        self.assertEqual(data["id"], product.id)
+        self.assertIn("name", data)
+        self.assertEqual(data["name"], product.name)
+
+
+    def test_deserialize_a_product(self):
+        """ Test deserialization of a Product """
+        data = {
+            "id": 1,
+            "name": "Macbook"
+        }
+        product = Product()
+        product.deserialize(data)
+        self.assertNotEqual(product, None)
+        self.assertEqual(product.id, None)
+        self.assertEqual(product.name, "Macbook")
+
+
+    ######################################################################
+    #  A S S O C I A T I O N   T E S T   C A S E S
+    ######################################################################
+        
+    # def test_create_a_supplier_with_product(self):
```

Comment:
Lets uncomment this and add the missing tests for the association

Reply:
Added additional tests